modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/utils/modules.js

```diff
@@ -14,14 +14,76 @@
 
 import React from 'react';
 
+const moduleURLPromises = {};
+
 export function getLiferayJsModule(moduleURL) {
-	return new Promise((resolve, reject) => {
-		Liferay.Loader.require(
-			moduleURL,
-			(jsModule) => resolve(jsModule.default || jsModule),
-			(error) => reject(error)
-		);
-	});
+	if (moduleURL.includes('|')) {
+		if (!moduleURLPromises[moduleURL]) {
+			moduleURLPromises[moduleURL] = new Promise((resolve, reject) => {
+				// GreenAppleFDSCellRenderer|/o/ce-sample/index.js
+				const parts = moduleURL.split('|');
+
+				// TODO: We may want to create a npm-scripts webpack plugin to add /*webpackIgnore: true*/ to every dynamic import
+				import(/*webpackIgnore: true*/ parts[1])
+					.then((module) => {
+						let renderer = module[parts[0]];
+
+						resolve((props) => {
+							return React.createElement(
+								class extends React.Component {
```

Comment:
Can the props of the cell change dynamically on a re-render of the react tree? I don't think it can... And if it can't we may consider adding a `shouldUpdate` and set it to false so that we don't get unnecessary re-renders as it might be expensive for whatever the customer provides.

Reply:
The rendering lifecyle is one of my main concerns because it's part of the contract, but can only be expressed with words (we cannot create a Typescript type to force it) so we'd better define it correctly and explicitly and make it evolvable and maintainable if we don't want to have problems in the future.

The current lifecycle seemed a bit flawed to me bc I saw some runtime smells like, for example, infinite loops if the renderer was not resolved synchronously or on the next tick :grimacing:. We will need to investigate it and define what we really want. @markocikos should be involved in this part too, I guess.

Reply:
I think if we ensure that the return content of the renderer is a string, we don't need to then create this custom react code here and have a lifecycle, we can simply just return the string content to where the cell is rendered.